### PR TITLE
use consignment transfer to implement reclaiming bucket

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: action/consignment_transfer.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	"encoding/json"
+
+	"github.com/ethereum/go-ethereum/accounts"
+
+	"github.com/iotexproject/go-pkgs/crypto"
+	"github.com/iotexproject/iotex-address/address"
+)
+
+type (
+	// a consignment is a transaction signed by transferee which contains an embedded message signed by transferor
```
Comment (dustinxie) [inline action/consignment_transfer.go:20]: see https://etherscan.io/verifySig/2070 for an example

File: action/consignment_transfer.go [hunk 2]
```diff
@@ -46,18 +49,25 @@ type (
 
 	// ConsignMsg is the consignment message
 	ConsignMsg struct {
-		Index     int    `json:"index"`
+		Type      string `json:"type"`
```
Comment (dustinxie) [inline action/consignment_transfer.go:52]: add type of signature/chain, per Zhi's comment 1

File: action/consignment_transfer.go [hunk 3]
```diff
@@ -46,18 +49,25 @@ type (
...
+		Index     int    `json:"bucket"`
 		Nonce     int    `json:"nonce"`
 		Recipient string `json:"recipient"`
-		Comment   string `json:"comment"`
+		Reclaim   string `json:"reclaim"`
```
Comment (dustinxie) [inline action/consignment_transfer.go:56]: checked with Larry, use "reclaim"

File: action/consignment_transfer.go [hunk 4]
```diff
@@ -46,18 +49,25 @@ type (
...
+	}
+
+	// ConsignJSON is the JSON format of a consignment, it contains a message (which is ConsignMsg), and a signature
+	ConsignJSON struct {
```
Comment (dustinxie) [inline action/consignment_transfer.go:60]: Zhi's comment 3

File: action/consignment_transfer.go [hunk 5]
```diff
@@ -102,30 +118,35 @@ func (c *consignment) Transferee() address.Address {
...
 	return nil, crypto.ErrInvalidKey
 }
 
-func msgHash(msg []byte) [][]byte {
-	h1, _ := accounts.TextAndHash(msg) // Ethereum's way to calculate hash for the given message
-	// TODO: add hash calculation for other ECC algorithms or coins
-	return [][]byte{h1}
+func msgHash(sigType string, msg []byte) ([]byte, error) {
+	switch sigType {
+	case "Ethereum":
```
Comment (dustinxie) [inline action/consignment_transfer.go:146]: check sig/chain type per comment 1

File: action/protocol/staking/handlers.go [hunk 6]
```diff
@@ -318,13 +318,12 @@ func (p *Protocol) handleTransferStake(ctx context.Context, act *action.Transfer
...
 	log.AddTopics(byteutil.Uint64ToBytesBigEndian(bucket.Index), act.VoterAddress().Bytes(), bucket.Candidate.Bytes())
 
-	// check if the payload contains a valid consignment transfer
 	newOwner := act.VoterAddress()
-	if consignment, ok := p.handleConsignmentTransfer(blkCtx, actionCtx, act, bucket); ok {
-		newOwner = consignment.Transferee()
-	} else {
-		// regular transfer, verify caller owns the bucket
-		if !address.Equal(bucket.Owner, actionCtx.Caller) {
+	if !address.Equal(bucket.Owner, actionCtx.Caller) {
```
Comment (dustinxie) [inline action/protocol/staking/handlers.go:322]: per Zhi's comment 2

File: action/consignment_transfer.go (hunk 2 above)
Comment (CoderZhi) [reply action/consignment_transfer.go:52]: "Type" is supposed to be a member of ConsignJSON.

File: action/consignment_transfer.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		Transferor() address.Address
+		Transferee() address.Address
+		AssetID() uint64
+		TransfereeNonce() uint64
+	}
+
+	// ConsignMsg is the consignment message
+	ConsignMsg struct {
+		Type      string `json:"type"`
+		Index     int    `json:"bucket"`
```
Comment (CoderZhi) [inline action/consignment_transfer.go:53]: normally, the name and the json name should be consistent.

File: action/consignment_transfer.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	// ConsignJSON is the JSON format of a consignment, it contains a message (which is ConsignMsg), and a signature
+	ConsignJSON struct {
+		Msg string `json:"msg"`
+		Sig string `json:"sig"`
+	}
+
+	consignment struct {
+		ConsignJSON
```
Comment (CoderZhi) [inline action/consignment_transfer.go:66]: ConsignJSON does not need to be part of consignment.

File: action/consignment_transfer.go (hunk 7 above)
Comment (yutongp) [reply action/consignment_transfer.go:53]: maybe bucketIdx

**Pull request merged**